Hijack support

iconara opened the pull request:
This is an attempt at implementing (partial) `rack.hijack` support, as a follow up to #15. It removes support for `async.callback`, since it doesn't seem to work in Fishwife currently, and it also seems like it was abandoned in Rack in favour of `rack.hijack`.

Fishwife will probably not support all of the `rack.hijack` spec, it doesn't look like it's possible to get hold of the raw output stream without Jetty writing the HTTP header.

iconara committed:
Remove support for async.callback

It doesn't work in its current form and will be replaced.

iconara committed:
Implement support for response hijacking

When the response contains a `rack.hijack` header the response is handled asynchronously by handing off an IO like object to the callback in that header (after first committing the other response headers).

This is only half of the rack.hijack spec, and it might be the only part that is possible to support – the request part requires total control over the output, not even headers should be sent. Unfortunately there is no way to tell the application which part is supported so this sets the `rack.hijack?` property to true, and sets `rack.hijack` to a proc that raises NotImplementedError. It's not pretty, but at least it passes Rack::Lint.

iconara commented:
This is WIP, the test for the feature is not good enough.

iconara committed:
Remove support for #write_nonblock on hijacked IO

It was not like it was going to be very useful anyway.

iconara committed:
Write raw bytes instead of converting through Java strings

iconara committed:
Fix quoting in error messages

dekellum commented:
Thanks for PR. Moving comment to here...

1. I don't think there any problem using a servlet 3.1 call, which has been bundled with rjack-jetty as of 9.1.0.0 which is below the minimum for current fishwife.

2. See (java) [IOUtil.writeBody](https://github.com/dekellum/fishwife/blob/fishwife-dev/src/main/java/fishwife/IOUtil.java#L91) for a potentially better way to get a (ruby) String written to OutputStream with no additional copies or char conversion.

iconara commented:
Looks like I could do `IOUtil.write_body([str], @async_context.response.output_stream)`, but that would instead create a new array for each call instead. Would you accept `@async_context.response.output_stream.write(str.to_java_bytes)`? Otherwise I could write a version of `IOUtil.write_body` that writes just a single string and not a complete body.

dekellum commented:
I could accept to_java_bytes, but then I might want to do the IOUtil conversion before a release. If you are setup well enough for java-dev and want to try the IOUtil change yourself, the particulars are in AppendBlockCallback.call(). Might be able to just move that impl. for reuse. 

I'm also kind of surprised that the rack.hijack interface doesn't also support arrays though, as it would be symmetric with the blocking interface.  I suppose HijackIO#write could just do the right thing if passed an array, analogous to IO#puts , and really just an optimization of `IO#write( obj ) {  obj = obj.to_s }`

dekellum commented:
I made a mistake on the last point:  IO#write( obj ) interface is indeed defined to call obj.to_s, so HijackIo#write should also do that, at least for all other types.  However, I didn't recall that Array#to_s is seemingly arbitrarily just an alias for Array#inspect

~~~ ruby
['a', 'b'].to_s # -> ["a", "b"]
~~~

It seems that would be rather undesirable in this context.  Could add a HijackIo#write_array, or the like at least to document the behavior and for convenience.

iconara committed:
Add IOUtil.write

Writes a single string to an output stream.

iconara committed:
Use IOUtil.write in HijackIo

iconara committed:
Extract the common code between IOUtil.write and .writeBody to its own method

iconara committed:
Flush the output stream, not the writer

It probably accomplishes the same thing, but it's inconsistent (#write used the writer in my first spike, and I seem to have forgotten to change #flush when I rewrote #write).

iconara committed:
Add proper tests for HijackIo and make it work more like an actual IO

iconara commented:
I haven't had much time to work on this lately, but I've now added Java code for writing to the output stream, and tests for `HijackIo`.
